crates/sui-benchmark/tests/simtest.rs

```diff
@@ -249,10 +257,129 @@ mod test {
+        // way to the latest protocol version.
+        let min_ver = ProtocolVersion::MIN.as_u64();
+        let max_ver = ProtocolVersion::MAX.as_u64();
+        let init_framework =
+            sui_framework_snapshot::load_bytecode_snapshot("testnet", min_ver).unwrap();
+        let mut test_cluster = init_test_cluster_builder(7, 3000)
+            .with_protocol_version(ProtocolVersion::MIN)
+            .with_supported_protocol_versions(SupportedProtocolVersions::new_for_testing(
+                min_ver, min_ver,
+            ))
+            .with_fullnode_supported_protocol_versions_config(
+                SupportedProtocolVersions::new_for_testing(min_ver, max_ver),
+            )
+            .with_objects(init_framework)
+            .build()
+            .await
+            .unwrap();
+        //let test_init_data = TestInitData::new(&test_cluster).await;
+
+        // It is known that protocol v1 has a bug that could lead to safe mode.
+        test_cluster.set_safe_mode_expected(true);
+
+        let finished = Arc::new(AtomicBool::new(false));
+        let finished_clone = finished.clone();
+        let _handle = tokio::task::spawn(async move {
+            for version in min_ver..max_ver {
+                info!("Current version: {}", version);
+                tokio::time::sleep(Duration::from_secs(20)).await;
+                test_cluster.all_node_handles().for_each(|h| {
+                    h.with(|node| {
+                        assert_eq!(
+                            node.state()
+                                .get_sui_system_state_object_for_testing()
+                                .unwrap()
+                                .protocol_version(),
+                            version
+                        )
+                    })
+                });
+                let next_version = version + 1;
+                let new_framework =
+                    sui_framework_snapshot::load_bytecode_snapshot("testnet", next_version);
+                let new_framework_modules: Vec<_> = match new_framework {
+                    Ok(new_framework) => new_framework
+                        .into_iter()
+                        .flat_map(|object| {
+                            object
+                                .data
+                                .try_as_package()
+                                .unwrap()
+                                .deserialize_all_modules()
+                                .unwrap()
+                        })
+                        .collect(),
+                    Err(_) => {
+                        assert_eq!(next_version, max_ver);
+                        sui_framework::make_system_modules()
+                            .into_iter()
+                            .flatten()
+                            .collect()
+                    }
+                };
+                sui_system_injection::set_override(new_framework_modules);
```

Comment:
This function only overrides the `sui_system` package, but `new_framework_modules` contains modules from all the system packages.  If you want to use this system, you'll want to create override mechanisms for all three system packages.

At that point, it might be most convenient to abstract the test injection logic into the `define_system_package` macro:

https://github.com/MystenLabs/sui/blob/sui-upgrade-simtests/crates/sui-framework/src/lib.rs#L29